mlir/lib/Quantum/Transforms/DisentangleCNOT.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This algorithm is taken from https://arxiv.org/pdf/2012.07711, table 1
+
+#define DEBUG_TYPE "disentanglecnot"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/Support/Debug.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+
+#include "PropagateSimpleStatesAnalysis.hpp"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+#define GEN_PASS_DEF_DISENTANGLECNOTPASS
+#define GEN_PASS_DECL_DISENTANGLECNOTPASS
+#include "Quantum/Transforms/Passes.h.inc"
+
+struct DisentangleCNOTPass : public impl::DisentangleCNOTPassBase<DisentangleCNOTPass> {
+    using impl::DisentangleCNOTPassBase<DisentangleCNOTPass>::DisentangleCNOTPassBase;
+
+    bool canScheduleOn(RegisteredOperationName opInfo) const override
+    {
+        return opInfo.hasInterface<FunctionOpInterface>();
+    }
+
+    void runOnOperation() override
+    {
+        func::FuncOp func = cast<func::FuncOp>(getOperation());
+        mlir::IRRewriter builder(func->getContext());
+        Location loc = func->getLoc();
+
+        PropagateSimpleStatesAnalysis &pssa = getAnalysis<PropagateSimpleStatesAnalysis>();
+        llvm::DenseMap<Value, QubitState> qubitValues = pssa.getQubitValues();
+
+        if (EmitFSMStateRemark) {
+            for (auto it = qubitValues.begin(); it != qubitValues.end(); ++it) {
+                it->first.getDefiningOp()->emitRemark(pssa.QubitState2String(it->second));
+            }
+        }
+
+        func->walk([&](quantum::CustomOp op) {
```

Comment:
This should be a pattern instead of just a walk, like the other peephole passes. I don't know if I have enough time, but let me try to convert it to a pattern.

Reply:
It really depends, not everything is suited for pattern rewriting. Is the CNOT disentangling something local, that would benefit from being applied independently and in an iterative fashion across the IR?

Reply:
So the benefit of having it as a pattern is that we have applicators that can repeatedly apply until a fixed point. This would help cases like 
```
qml.CNOT(0, 1)
qml.CNOT(0, 1)
```

For the "just a single walk", only the first CNOT will get removed, since the qubit value on the first wire behind the first CNOT is not obviously |0> yet in the first (and only) run of the analysis, as its parent is a CNOT, which is not an entry in the FSM.

The downside is of course multiple runs of the analysis which will likely be mostly the same. This can be optimized by caching the analysis object (or at least the map inside), and only updating the analysis across different pattern applications instead of restarting the analysis from scratch. Although, this would definitely take more than a day.

We don't have to change it to a pattern (since likely I won't have enough time now). We can leave this comment unresolved if anyone wants to try changing it and see the impacts.

Reply:
Yeah without updating the analysis there wouldn't be much point to it, but it would require a more complex system (ideally you can efficiently update a branch of the circuit and it would only re-run the analysis on that branch). I would recommend leaving as is.